### PR TITLE
feat: check connection availability before fetching sourceConfig

### PR DESCRIPTION
**Pull request by 1abhishekpandey**
## Description

- In this PR, we’ve added logic to check and wait for the connection to be available before attempting to fetch the sourceConfig. This ensures that the sourceConfig is not fetched until the network is available.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactor/optimization

### SourceConfigManager

- First, we will fetch the cached sourceConfig from storage, if available, and immediately notify the observers with the cached sourceConfig.
- Then, we will refresh the sourceConfig by downloading the latest version, store it, and notify all observers.
- Added test cases: Rewrote them from scratch.

### Core Analytics

- Refactored the setup portion of the core Analytics

### LoggerAnalytics

- Improved LoggerAnalytics by handling edge cases. I noticed that if it’s not initialized early in the lifecycle and a message is logged, it throws a fatal exception. Now, in such cases, we simply ignore any logging requests.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [ ] I have ensured that my code follows the project's code style.
- [ ] I have checked for potential performance impacts and optimized if necessary.
- [ ] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?

- Run the Android app and toggle on the connection. The SC will be immediately fetched, as soon as the connection becomes available.
- Run the Android app on different SDK API versions i.e., compatible SDK version - API 24 and above and legacy once API 21 - 23.
- Run the core module, and an attempt will be made to fetch the sourceConfig immediately, as we assume (or the default behaviour is) that the network is always available in core SDKs.
- Try to enable and disable connection multiple times, but sourceConfig will only be fetched once.

Log:
```
SourceConfig is fetched successfully: SourceConfig(source=RudderServerConfigSource(sourceId=......
```

## Maintainers Checklist

- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

### CODE REVIEW

**Review by ChryssaAliferi — APPROVED** [review]
Great effort!! 🚀

File: core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ import org.jetbrains.annotations.VisibleForTesting
  *
  * @property source The configuration details of a RudderStack source.
  */
+@Suppress("MaximumLineLength")
```
Comment (vgupta98) [inline core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt:18]: NIT: We don't need to suppress `MaximumLineLength` rule here. We can have `updatedSourceConfig` of `NotifyObserversAction` on a new line. For some reason, detekt fails to autocorrect it.

File: core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/utils/ConnectivityUtils.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * It notifies the observers when the connection is available.
+ */
+@InternalRudderApi
+fun Analytics.notifyOnlyOnceOnConnectionAvailable(block: suspend () -> Unit) {
+    this.analyticsScope.launch {
+        this@notifyOnlyOnceOnConnectionAvailable.connectivityState
+            .filter { it }
+            .take(ONLY_ONE_ELEMENT)
+            .collect { block() }
+    }
```
Comment (vgupta98) [inline core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/utils/ConnectivityUtils.kt:21]: Nice!

File: core/src/test/kotlin/com/rudderstack/sdk/kotlin/core/SourceConfigManagerTest.kt [hunk 3]
```diff
@@ -35,9 +43,20 @@ class SourceConfigManagerTest {
...
     private lateinit var sourceConfigManager: SourceConfigManager
 
+    @MockK
+    private lateinit var mockFlowState: FlowState<Boolean>
+
+    private lateinit var flowCollectorSlot: CapturingSlot<FlowCollector<Boolean>>
+
     @Before
     fun setUp() {
-        Dispatchers.setMain(testDispatcher)
```
Comment (vgupta98) [inline core/src/test/kotlin/com/rudderstack/sdk/kotlin/core/SourceConfigManagerTest.kt:40]: Remove the Dispatchers.resetMain() from teardown as well.

File: core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt [hunk 4]
```diff
@@ -34,7 +35,7 @@ internal data class SourceConfig(
         )
     }
 
-    class UpdateAction(@VisibleForTesting internal val updatedSourceConfig: SourceConfig) : FlowAction<SourceConfig> {
+    class NotifyObserversAction(@VisibleForTesting internal val updatedSourceConfig: SourceConfig) : FlowAction<SourceConfig> {
```
Comment (vgupta98) [inline core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt:38]: Suggestion: IMO we should not call it `NotifyObserversAction` because notifying the observers is the inherent property of a `MutableStateFlow` and it is not the responsibility of this specific action. This action's responsibility is to just update (or set) the sourceConfig state. Thoughts?

File: core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/Analytics.kt [hunk 5]
```diff
@@ -78,11 +79,30 @@ open class Analytics protected constructor(
...
         setup()
         storeAnonymousId()
     }
 
+    private fun runForBaseTypeOnly() {
+        if (this::class == Analytics::class) {
+            setLogger(logger = KotlinLogger())
+            connectivityState.dispatch(ConnectivityState.SetDefaultStateAction())
+            setupSourceConfig()
+        }
```
Comment (vgupta98) [inline core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/Analytics.kt:93]: Great job in detecting this subtle error about base and parent class!

File: core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt (hunk 4 above)
Comment (1abhishekpandey) [reply core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/models/SourceConfig.kt]: Agree. Thanks for the clarification.

**Review by vgupta98 — APPROVED** [review]
